src/librustc_passes/rvalue_promotion.rs

```diff
@@ -420,8 +427,9 @@ fn check_expr_kind<'a, 'tcx>(
                     // Just in case the type is more specific than the definition,
                     // e.g. impl associated const with type parameters, check it.
                     // Also, trait associated consts are relaxed by this.
-                    if !(promotable || v.type_has_only_promotable_values(node_ty)) {
-                        return NotPromotable
+                    let promotability_output = v.type_promotability(node_ty);
+                    if !promotable {
+                        return promotability_output;
                     }
                     Promotable
```

Comment:
remove this and just place the `v.type_promotability(node_ty)` call here and let the value fall through